pkg/k8s/watchers/node.go

```diff
@@ -35,6 +36,13 @@ func (k *K8sWatcher) nodesInit(k8sClient kubernetes.Interface) {
 		&slim_corev1.Node{},
 		0,
 		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if node := k8s.ObjToV1Node(obj); node != nil {
+					if option.Config.BGPAnnounceLBIP {
+						k.bgpSpeakerManager.OnUpdateNode(node) // Need to seed the BGP speaker
+					}
+				}
+			},
```

Comment:
:thinking: why did we never subscribe to these events before? Seems a bit odd.

Reply:
Good question, :shrug:

Reply:
cc @aanm do you happen to know why we didn't previously care about Node Add events?

Reply:
I'll defer this to @pchaigno since this watcher was only used labels in the host endpoint. @pchaigno How is the host endpoint getting the labels from a k8s node when the agent starts?

Reply:
Oh I see, that makes sense. We wouldn't ever see an Add event for the local node because it would have to be provisioned before Cilium can run on it.

Reply:
^^ @christarazi does that mean we don't actually need this snippet? Seems like we're only listening for node updates for the local node (`fields.ParseSelectorOrDie("metadata.name="+nodeTypes.GetName())`)

Reply:
> @pchaigno How is the host endpoint getting the labels from a k8s node when the agent starts?

My understanding was the same as Joe's: ADDs can't happen because we're only watching for updates on the local node.

Reply:
@joestringer This snippet is needed to "seed" the MetalLB logic with the node labels which are needed to decide whether a node should be considered for peering. Essentially, this allows you to designate which nodes should connect to the BGP router(s) and ultimately share routes to. I've found that when the K8s node watcher is initialized, it'll notify with an add event, and not with an update event which the BGP code is already hooked up with. Basically, what hooking into an add event gives us is a signal of when the node labels are available to seed MetalLB with. This is probably not the only way to do this, but I'm not sure of another part of Cilium where we care for the node labels. However, hooking into the K8s watcher seemed like a natural place to do this especially because BGP support depends on the K8s watcher already anyway.

Reply:
> I've found that when the K8s node watcher is initialized, it'll trigger with an add event, and not with an update event

That's a super interesting observation and suggests that we should be handling this event for the hostfw case as well.

At a high level my question is "Is there a bug here that we should prepare a fix for and backport that?".

Reply:
Stepping back a bit, I think we do have a bug. It seems that we mis-assumed that it was impossible to receive a node add event for the local node. That actually doesn't make much sense if we think about it. The K8s watcher is essentially a cache of the K8s resources in Cilium. When Cilium is coming up, it'll initialize this cache. When the informers (watchers) first come up, they'll sync up with the apiserver and populate the local cache. During this, each new resource seen will  trigger an add event for each resource, and in this case, an add event for the local node.

cc @aanm to double check me on this

Reply:
> During this, each new resource seen will trigger an add event for each resource, and in this case, an add event for the local node.

correct.

Reply:
> > I've found that when the K8s node watcher is initialized, it'll trigger with an add event, and not with an update event
> 
> That's a super interesting observation and suggests that we should be handling this event for the hostfw case as well.
> 
> At a high level my question is "Is there a bug here that we should prepare a fix for and backport that?".


It looks like this could be the cause for bugs https://github.com/cilium/cilium/issues/13676 and/or https://github.com/cilium/cilium/issues/13455. I'll send a pull request, hopefully today.